staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache_interval.go

```diff
@@ -0,0 +1,223 @@
+	indexer indexerFunc
+
+	// indexValidator is used to check if a given
+	// index is still valid perspective. If it is
+	// deemed that the index is not valid, then this
+	// interval can no longer be used to serve events.
+	//
+	// An interval is invalid when its startIndex
+	// no longer coincides with the startIndex of
+	// underlying watchCache and therefore does not
+	// have meaning for cases where the events are
+	// served from the underlying store instead of
+	// the watchCache (rv = 0).
+	indexValidator indexValidator
+
+	// buffer holds watchCacheEvents that this interval
+	// returns on a call to Next(). This exists mainly
+	// to reduce acquiring the lock on each invocation
+	// of Next().
+	buffer *watchCacheIntervalBuffer
+
+	// lock effectively guards the valid field.
+	lock sync.Locker
+}
+
+type attrFunc func(runtime.Object) (labels.Set, fields.Set, error)
+type indexerFunc func(int) *watchCacheEvent
+type indexValidator func(int) bool
+
+func newCacheInterval(startIndex, endIndex int, indexer indexerFunc, indexValidator indexValidator, locker sync.Locker) *watchCacheInterval {
+	return &watchCacheInterval{
+		startIndex:     startIndex,
+		endIndex:       endIndex,
+		indexer:        indexer,
+		indexValidator: indexValidator,
+		buffer:         &watchCacheIntervalBuffer{buffer: make([]*watchCacheEvent, bufferSize)},
+		lock:           locker,
+	}
+}
+
+// newCacheIntervalFromStore is meant to handle the case of rv=0, such that the events
+// returned by Next() need to be events from a List() done on the underlying store of
+// the watch cache.
+func newCacheIntervalFromStore(resourceVersion uint64, store cache.Indexer, getAttrsFunc attrFunc, locker sync.Locker) (*watchCacheInterval, error) {
+	buffer := &watchCacheIntervalBuffer{}
+	allItems := store.List()
+	buffer.buffer = make([]*watchCacheEvent, len(allItems))
+	for i, item := range allItems {
+		elem, ok := item.(*storeElement)
+		if !ok {
+			return nil, fmt.Errorf("not a storeElement: %v", elem)
+		}
+		objLabels, objFields, err := getAttrsFunc(elem.Object)
+		if err != nil {
+			return nil, err
+		}
+		buffer.buffer[i] = &watchCacheEvent{
+			Type:            watch.Added,
+			Object:          elem.Object,
+			ObjLabels:       objLabels,
+			ObjFields:       objFields,
+			Key:             elem.Key,
+			ResourceVersion: resourceVersion,
+		}
+		buffer.endIndex++
+	}
+	ci := &watchCacheInterval{
+		startIndex: 0,
+		// Simulate that we already have all the events we're looking for.
+		endIndex: 0,
+		buffer:   buffer,
+		lock:     locker,
+	}
+
+	return ci, nil
+}
+
+// Next returns the next item in the cache interval provided the cache
+// interval is still valid. An error is returned if the interval is
+// invalidated.
+func (wci *watchCacheInterval) Next() (*watchCacheEvent, error) {
+	// if there are items in the buffer to return, return from
+	// the buffer.
+	if event, exists := wci.buffer.next(); exists {
+		return event, nil
+	}
```

Comment:
@wojtek-t while updating the tests, I realized one thing - we are using an interval buffer here to avoid frequent acquire-release of the lock, and as a result, we check here if the interval buffer has elements and return from that if it does. However, in the meantime, the interval itself could have been invalidated and we don't do a check for that while returning the element from the buffer. 

- Am I misunderstanding something here and is the check not required in this case?
- If what I said makes sense then we again get the issue of frequent acquire and release of lock because of the indexValidator?

Reply:
> Am I misunderstanding something here and is the check not required in this case?

Yes. The fact you're missing is that the pattern we're using here is:

1. items sit in original watchcache
2. we check if the ones we need are valid and copy them to our interval
3. now, we have all the data we need - we no longer need the data to sit in the watchcache - it can be overwritten, because we already copied it

So we only need to check for if the data we need to read from watchcache are valid - once we copied them to our buffer, we already know that these are the data we actually need

Reply:
Ah right, thanks for that clarification!